pg_replicate/src/clients/postgres.rs

```diff
@@ -136,13 +136,37 @@ impl ReplicationClient {
         Ok(stream)
     }
 
-    /// Returns a vector of columns of a table
+    /// Returns a vector of columns of a table, optionally filtered by a publication's column list
     pub async fn get_column_schemas(
         &self,
         table_id: TableId,
+        publication: Option<&str>,
     ) -> Result<Vec<ColumnSchema>, ReplicationClientError> {
+        let (pub_cte, pub_pred) = if let Some(publication) = publication {
+            (
+                format!(
+                    "with pub_attrs as (
+                        select unnest(r.prattrs)
+                        from pg_publication_rel r
+                        left join pg_publication p on r.prpubid = p.oid
+                        where p.pubname = '{publication}'
+                        and r.prrelid = {table_id}
+                    )"
+                ),
+                "and (
+                    case (select count(*) from pub_attrs)
+                    when 0 then true
+                    else (a.attnum in (select * from pub_attrs))
+                    end
+                )",
+            )
+        } else {
+            ("".into(), "")
+        };
```

Comment:
Does the new fix work when published columns skip over an unpublished column? E.g. attnum 1 and 3 are published but attnum 2 is not.

Reply:
This query returns the right thing in that case yes but I'll go do a full test with data to be sure.

Reply:
Okay, this _does_ work, but I've hit on a different unrelated error while testing which has revealed a missing bit in this PR wrt the initial table copy logic.